outline/shadowsocks/client.go

```diff
@@ -22,36 +27,70 @@ import (
 
 	"github.com/Jigsaw-Code/outline-go-tun2socks/outline"
 	"github.com/Jigsaw-Code/outline-go-tun2socks/outline/connectivity"
+	"github.com/Jigsaw-Code/outline-go-tun2socks/outline/internal/utf8"
 	"github.com/Jigsaw-Code/outline-go-tun2socks/outline/neterrors"
 	"github.com/Jigsaw-Code/outline-internal-sdk/transport"
 	"github.com/Jigsaw-Code/outline-internal-sdk/transport/shadowsocks"
 	"github.com/Jigsaw-Code/outline-internal-sdk/transport/shadowsocks/client"
 	"github.com/eycorsican/go-tun2socks/common/log"
 )
 
-// Config represents a shadowsocks server configuration.
-// Exported via gobind.
-type Config struct {
-	Host       string
-	Port       int
-	Password   string
-	CipherName string
-	Prefix     []byte
-}
-
+// A client object that can be used to connect to a remote Shadowsocks proxy.
 type Client = outline.Client
 
-// NewClient provides a gobind-compatible wrapper for [client.NewClient].
+// NewClient creates a new Shadowsocks client from a non-nil configuration.
+//
+// [Deprecated] Please use NewClientFromJSON or NewClientFromParameters.
 func NewClient(config *Config) (*Client, error) {
+	if config == nil {
+		return nil, fmt.Errorf("Shadowsocks configuration is required")
+	}
+	return newClient(config.Host, config.Port, config.CipherName, config.Password, config.Prefix)
+}
+
+// NewClientFromJSON creates a new Shadowsocks client from a JSON formatted configuration.
+func NewClientFromJSON(configJSON string) (*Client, error) {
+	config, err := parseConfigFromJSON(configJSON)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to parse Shadowsocks configuration JSON: %w", err)
+	}
+	return NewClientFromParameters(config.Host, int(config.Port), config.Method, config.Password, config.Prefix)
+}
+
+// NewClientFromParameters creates a new Shadowsocks client from the parameters.
+//
+//   - host specifies the hostname/IP of the Shadowsocks server
+//   - port specifies the port number of the Shadowsocks server
+//   - cipher and password specifies the encryption strategy
+//   - prefix is an optional string, it specifies the salt prefix used in the
+//     beginning of the session. The should be a UTF-8 encoded string with each
+//     codepoint representing a single byte (0x00 ~ 0xFF) in the salt prefix.
+func NewClientFromParameters(host string, port int, cipher, password string, prefix string) (*Client, error) {
```

Comment:
The difference is this function uses `prefix` as a string, but the legacy `NewClient` uses it as a `byte[]`. I'm trying to prevent the decode process happening from multiple places.

Reply:
Removed the new function, adding a private function to `main.go` instead.